model/src/generator.cpp

```diff
@@ -53,5 +53,34 @@ const tile* tilePermutator::generate() {
 	return result;
 }
 
+historyRoll::historyRoll(
+	const tile* tiles[], std::size_t numTiles,
+	std::size_t retryTimes, std::size_t* initialHistory,
+	std::size_t historySize, uint64_t seed):
+	history(initialHistory, initialHistory + historySize),
+	tiles(new const tile*[numTiles]), numTiles(numTiles),
+	retryTimes(retryTimes), historySize(historySize),
+	randomizer(seed) {}
+
+const tile* historyRoll::generate() {
+	std::uniform_int_distribution<std::size_t> sampler{0, numTiles - 1};
+	std::size_t result = sampler(randomizer);
+	for (std::size_t i = 0; i < retryTimes; ++i) {
+		// TODO: better way to find previously generated tiles
```

Comment:
It seems that a deque plus an array of length `numTiles` storing the reference count will do.

Reply:
When a tile is enqueued, the reference counter increments, and vice versa.

Reply:
fixed by c76f353